contracts/core/PositionVault.sol

```diff
@@ -316,43 +294,140 @@ contract PositionVault is Constants, ReentrancyGuard, IPositionVault {
     }
 
+    function executeOpenMarketOrder(uint256 _posId) public nonReentrant {
+        require(settingsManager.isManager(msg.sender), "You are not allowed to trigger");
+
+        Position memory position = positions[_posId];
+        Order storage order = orders[_posId];
+        require(order.size > 0, "order size is 0");
+        require(order.positionType == POSITION_MARKET, "not market order");
+        require(order.status == OrderStatus.PENDING, "not pending order");
+
+        settingsManager.updateFunding(position.indexToken);
+
+        uint256 fee = settingsManager.collectMarginFees(
+            position.owner,
+            position.indexToken,
+            position.isLong,
+            order.size
+        );
+        uint256 price = priceManager.getLastPrice(position.indexToken);
+        checkSlippage(position.isLong, order.lmtPrice, price);
+
+        _increasePosition(
+            position.owner,
+            position.indexToken,
+            order.collateral + fee,
+            order.size,
+            lastPosId,
+            price,
+            position.isLong
+        );
+
+        order.collateral = 0;
+        order.size = 0;
+        order.status = OrderStatus.FILLED;
+        _addUserAlivePosition(position.owner, lastPosId);
+
+        emit UpdateOrder(_posId, order.positionType, order.status);
+    }
+
+    function _executeOpenMarketOrders(uint256 numOfOrders) internal {
+        uint256 index = openMarketQueueIndex;
+        uint256 endIndex = index + numOfOrders;
+        uint256 length = openMarketQueuePosIds.length;
+
+        if (index >= length) return;
+        if (endIndex > length) endIndex = length;
+
+        while (index < endIndex) {
+            uint256 posId = openMarketQueuePosIds[index];
+
+            try this.executeOpenMarketOrder(posId) {} catch Error(string memory err) {
+                cancelMarketOrder(posId);
+                emit MarketOrderExecutionError(posId, positions[posId].owner, err);
+            } catch (bytes memory) {
+                cancelMarketOrder(posId);
+            }
+
+            delete openMarketQueuePosIds[posId];
+            ++index;
+        }
+
+        openMarketQueueIndex = index;
+    }
+
+    function executeOpenMarketOrdersWithPrices(
+        uint256 numOfOrders,
+        address[] calldata indexTokens,
+        uint256[] calldata prices
+    ) external {
+        require(settingsManager.isManager(msg.sender), "You are not allowed to trigger");
+
+        priceManager.setLatestPrices(indexTokens, prices);
+        _executeOpenMarketOrders(numOfOrders);
+    }
+
+    function cancelMarketOrder(uint256 _posId) internal {
+        Order storage order = orders[_posId];
+        order.status = OrderStatus.CANCELED;
+        emit UpdateOrder(_posId, order.positionType, order.status);
+    }
+
+    function hasUnexecutedMarketOrders() external view returns (bool) {
+        return openMarketQueueIndex < openMarketQueuePosIds.length;
```

Comment:
maybe return `openMarketQueuePosIds.length-openMarketQueueIndex` instead of a bool

Reply:
returning `openMarketQueuePosIds.length-openMarketQueueIndex` will get the number of pending orders, it works too

Reply:
consider push a new commit to change this